### DISCUSSION TITLE
sort order of arrays

### DISCUSSION THREAD

**Discussion by stephanecharette** [nlohmann/json#4327] [Q&A]
I understand from issues such as https://github.com/nlohmann/json/issues/727 that key-value pairs are sorted alphabetically since a std::map is used.

But what about arrays?  Are they stored in a std::vector which may be easily sorted?  I attempted to std::sort an array, and while this initially seemed to work, with some JSONs we're seeing asserts and C++ exceptions from deep within the nlohmann library.  This makes me think we should not be messing with the internals.

The code we had attempted to use was something like this:

    std::sort(j.begin(), j.end(),
        [](const JSON & lhs, const JSON & rhs) -> bool
        {
            return some_fancy_sort_order_thing(lhs, rhs);
        }

Most of the time this appeared to work.  But every once in a while this would do stuff like rhs would be null, we'd get an assert, or a C++ exception would be thrown.

Some of the asserts I saw earlier today which I still have on my screen:

    void nlohmann::json_abi_v3_11_3::detail::serializer<BasicJsonType>::dump(const BasicJsonType&, bool, bool, unsigned int, unsigned int) [with BasicJsonType = nlohmann::json_abi_v3_11_3::basic_json<>]: Assertion `false' failed.

    nlohmann::json_abi_v3_11_3::detail::type_error: [json.exception.type_error.305] cannot use operator[] with a string argument with null

    nlohmann::detail::type_error: [json.exception.type_error.305] cannot use operator[] with a string argument with number

Can someone confirm or deny that we can/cannot std::sort json arrays?

**Comment by nlohmann** [comment]
I can confirm it *should* work, but since you found a way where it does not, I would be happy to see a concrete example so I could debug into it.

**Reply by stephanecharette** [reply]
Thanks.  This is a complex system.  I'll see if I can come up with a simple example.
